sw/host/opentitanlib/src/spiflash/flash.rs

```diff
@@ -0,0 +1,282 @@
+            _ => (sfdp.jedec.sector[0].size, sfdp.jedec.sector[0].erase_opcode),
+        };
+        SpiFlash {
+            size: sfdp.jedec.density,
+            erase_size: erase_sz,
+            erase_opcode: erase_op,
+            program_size: SpiFlash::LEGACY_PAGE_SIZE,
+            address_mode: AddressMode::from(sfdp.jedec.address_modes),
+            sfdp: Some(sfdp),
+        }
+    }
+
+    /// Create a new `SpiFlash` instance by reading an SFDP table from the `spi` Target.
+    pub fn from_spi(spi: &dyn Target) -> Result<Self> {
+        let sfdp = SpiFlash::read_sfdp(spi)?;
+        Ok(SpiFlash::from_sfdp(sfdp))
+    }
+
+    /// Set the SPI flash addressing mode to either 3b or 4b mode.
+    pub fn set_address_mode(&mut self, spi: &dyn Target, mode: AddressMode) -> Result<()> {
+        let opcode = [match mode {
+            AddressMode::Mode3b => SpiFlash::EXIT_4B,
+            AddressMode::Mode4b => SpiFlash::ENTER_4B,
+        }];
+        spi.run_transaction(&mut [Transfer::Write(&opcode)])?;
+        self.address_mode = mode;
+        Ok(())
+    }
+
+    /// Automatically set the addressing mode based on the size of the SPI flash.
+    pub fn set_address_mode_auto(&mut self, spi: &dyn Target) -> Result<()> {
+        self.set_address_mode(
+            spi,
+            if self.size <= 16 * 1024 * 1024 {
+                AddressMode::Mode3b
+            } else {
+                AddressMode::Mode4b
+            },
+        )
+    }
+
+    /// Read into `buffer` from the SPI flash starting at `address`.
+    pub fn read(&self, spi: &dyn Target, mut address: u32, buffer: &mut [u8]) -> Result<()> {
+        let mut remain = buffer.len();
+        let mut chunkstart = 0usize;
+        while remain != 0 {
+            // Break the read up according to the maximum chunksize the backend can handle.
+            let chunk = std::cmp::min(spi.max_chunk_size(), remain);
+            let chunkend = chunkstart + chunk;
+            let op_addr = self.opcode_with_address(SpiFlash::READ, address)?;
+            spi.run_transaction(&mut [
+                Transfer::Write(&op_addr),
+                Transfer::Read(&mut buffer[chunkstart..chunkend]),
+            ])?;
+            address += chunk as u32;
+            chunkstart += chunk;
+            remain -= chunk;
+        }
+        Ok(())
+    }
+
+    /// Erase a segment of the SPI flash starting at `address` for `length` bytes.
+    /// The address and length must be sector aligned.
+    pub fn erase(&self, spi: &dyn Target, mut address: u32, length: u32) -> Result<()> {
+        if address % self.erase_size != 0 {
+            return Err(Error::BadEraseAddress(address, self.erase_size).into());
+        }
+        if length % self.erase_size != 0 {
+            return Err(Error::BadEraseLength(length, self.erase_size).into());
+        }
+        let end = address + length;
+        while address < end {
```

Comment:
`for addr in (address..end).step_by(self.erase_size)`

Reply:
Done